File: engine/src/main/java/com/s3bender/service/SignatureService.java
package com.s3bender.service;

import org.springframework.stereotype.Service;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.HexFormat;

/**
 * HMAC-SHA256 request signing, shared by header-based auth and presigned URLs.
 * See /_docs/auth-and-signing.md for the exact string-to-sign layout.
 */
@Service
public class SignatureService {

    private static final String HMAC_ALGO = "HmacSHA256";

    public String sign(String secretKey, String stringToSign) {
        try {
            Mac mac = Mac.getInstance(HMAC_ALGO);
            mac.init(new SecretKeySpec(secretKey.getBytes(StandardCharsets.UTF_8), HMAC_ALGO));
            byte[] raw = mac.doFinal(stringToSign.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(raw);
        } catch (Exception e) {
            throw new IllegalStateException("Failed to compute HMAC signature", e);
        }
    }

    /** Constant-time comparison to avoid leaking signature material via timing. */
    public boolean matches(String expectedHex, String providedHex) {
        if (expectedHex == null || providedHex == null) {
            return false;
        }
        byte[] a = expectedHex.getBytes(StandardCharsets.UTF_8);
        byte[] b = providedHex.getBytes(StandardCharsets.UTF_8);
        return MessageDigest.isEqual(a, b);
    }

    public String stringToSignForHeader(String method, String path, long timestampEpochSeconds) {
        return method + "\n" + path + "\n" + timestampEpochSeconds;
    }

    public String stringToSignForPresign(String method, String path, long expiresEpochSeconds) {
        return method + "\n" + path + "\n" + expiresEpochSeconds;
    }
}

File: engine/Services/SignatureService.cs
using System.Security.Cryptography;
using System.Text;

namespace S3Bender.Api.Services;

/// <summary>
/// HMAC-SHA256 request signing, shared by header-based auth and presigned URLs.
/// See /_docs/auth-and-signing.md for the exact string-to-sign layout.
/// </summary>
public class SignatureService
{
    public string Sign(string secretKey, string stringToSign)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secretKey));
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(stringToSign));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    /// <summary>Constant-time comparison to avoid leaking signature material via timing.</summary>
    public bool Matches(string? expectedHex, string? providedHex)
    {
        if (expectedHex is null || providedHex is null) return false;
        var a = Encoding.UTF8.GetBytes(expectedHex);
        var b = Encoding.UTF8.GetBytes(providedHex);
        return CryptographicOperations.FixedTimeEquals(a, b);
    }

    public string StringToSignForHeader(string method, string path, long timestampEpochSeconds) =>
        $"{method}\n{path}\n{timestampEpochSeconds}";

    public string StringToSignForPresign(string method, string path, long expiresEpochSeconds) =>
        $"{method}\n{path}\n{expiresEpochSeconds}";
}
